flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/ExecutionVertexVersioner.java

```diff
@@ -83,6 +83,15 @@ public Set<ExecutionVertexID> getUnmodifiedExecutionVertices(
                 .collect(Collectors.toSet());
     }
 
+    public Map<ExecutionVertexID, ExecutionVertexVersion> getExecutionVertexVersions(
```

Comment:
What's the difference between this method and `recordVertexModifications` ? Why do we need this?

Reply:
`recordVertexModifications` will increment the version, which should not happen when scheduling speculative attempts.